bpo-34903: For datetime.datetime.strptime(), the leading zero for some two-digit formats is optional.

mikegleen opened the pull request:
This pull request replaces #9747 which I am closing. Sorry, couldn't figure out how to cleanly update the original.

I am adding tests to datetimetester.py following discussion with Paul Ganssle. Please let me know if this causes a problem.


<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
bpo-NNNN: Summary of the changes made
```

Where: bpo-NNNN refers to the issue number in the https://bugs.python.org.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `master`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `master`.

-->


<!-- issue-number: [bpo-34903](https://bugs.python.org/issue34903) -->
https://bugs.python.org/issue34903
<!-- /issue-number -->

mikegleen committed:
Update rst file

mikegleen committed:
Add tests for single digit dates and times.

mikegleen committed:
Cosmetic.

mikegleen committed:
Add comment.

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

pganssle commented on Misc/NEWS.d/next/Documentation/2019-06-17-09-36-46.bpo-34903.r_wGRc.rst:
```diff
@@ -0,0 +1,3 @@
+For datetime.datetime.strptime(), the leading zero for some two-digit formats is optional.
```
You are encouraged to add a "Patch by" message to your NEWS entries, to give you the well-deserved credit for the patch.

And since I am suggesting that you do this anyway, I feel less bad about bikeshedding on the content of the message: I think it can be reworded to be a bit more concise:

"Documented that in :meth:`datetime.datetime.strptime`, the leading zero in some two-digit formats is optional. Patch by Mike Gleen."

pganssle commented on Lib/test/datetimetester.py:
```diff
@@ -2501,6 +2501,47 @@ def test_strptime(self):
+                              format=format,
+                              target=target):
+                newdate = strptime(string, format)
+                self.assertEqual(newdate.isoformat(timespec='seconds'), target,
```
Is there a particular reason you are converting the datetime to a string representation and then comparing that, rather than just comparing the datetimes directly?

It adds another thing that can fail that is unrelated to what you are testing here, so I think it would be best avoided unless there's something about datetime equality semantics that makes comparing the strings a better test.

Instead, I recommend something like:

```python
dt1 = self.theclass(2003, 1, 2, 4, 5, 6)
dt2 = self.theclass(2003, 2, 1)

inputs = [
    ('%d', '1/02/03 4:5:6', '%d/%m/%y %H:%M:%S', dt1),
    ('%m', '01/2/03 4:5:6', '%d/%m/%y %H:%M:%S', dt1),
    ...
    ('%V', '6/4/2003', '%u/%V/%G', dt2),
    ...
```

Then you can change the assertion to: `self.assertEqual(newdate, target, msg=reason)`.

pganssle commented on Lib/test/datetimetester.py:
```diff
@@ -2501,6 +2501,47 @@ def test_strptime(self):
+                newdate = strptime(string, format)
+                self.assertEqual(newdate.isoformat(timespec='seconds'), target,
+                                 msg=reason)
+        for testcase in inputs:
```
Using a closure for this seems unusual to me, did you have a particular reason for doing it this way? I can see for example that it does help to isolate the individual test cases by creating a new scope for each subtest, *but* I do find it a bit harder to read.

If you didn't have any particular reason, I'm inclined to say you should just move the body of `time_helper` into this for loop. If you had a particular reason or feel strongly about it, I'm OK with leaving the closure in place, but I think you can make the test more readable by changing the following things:

1. Move the `subtest` call out of the function, so that the function body isn't as deeply nested.

2. Rename `time_helper` to something like `run_strptime_test_case` or `assert_strptime`.

Then the loop would be something like this:

```python
for testcase in inputs:
    with self.subTest(reason=testcase[0]):
        run_strptime_test_case(*testcase)
```

pganssle reviewed (dismissed):
These changes look mostly pretty good, I've commented inline with some nitpicks.

One thing I don't want to get lost in the migration from #9747 is that Brett suggested that the tests be moved to `test_strptime`. I gave my reasoning why I think `datetimetester.py` [in this comment on the original issue](https://github.com/python/cpython/pull/9747#discussion_r294086307), but I would like @brettcannon to weigh in on this before just overriding his suggestion.

Finally, I should note that once the changes are approved, I'd like to run these tests against the buildbots *before* merging. `strptime` and `strftime` are particularly sensitive to weird edge cases on the "long tail" platforms, so changes like this are very likely to need adjustment after they've been run on the buildbots.

mikegleen committed:
Update 2019-06-17-09-36-46.bpo-34903.r_wGRc.rst

mikegleen committed:
Move the subtest logic in line with the for loop.
Replace the string used as a target with a datetime instance.

pganssle commented on Misc/NEWS.d/next/Documentation/2019-06-17-09-36-46.bpo-34903.r_wGRc.rst:
```diff
@@ -0,0 +1 @@
+Documented that in :meth:datetime.datetime.strptime(), the leading zero in some two-digit formats is optional. Patch by Mike Gleen.
```
```suggestion
Documented that in :meth:`datetime.datetime.strptime()`, the leading zero in some two-digit formats is optional. Patch by Mike Gleen.
```

This needs backticks so Sphinx can turn it into a link to the relevant method.

mikegleen committed:
Update 2019-06-17-09-36-46.bpo-34903.r_wGRc.rst

Add backticks so Sphinx can turn the method name into a link to the relevant method.

pganssle reviewed (approved):
I ran this through the buildbots yesterday and it did not create any new failures. The only semi-related failure was on [Alpine Linux](https://buildbot.python.org/all/#/builders/149/builds/43), but that is related to an earlier PR, so it seems like this documentation is accurate for all supporting platforms.

Thanks @mikegleen!

vstinner commented on Lib/test/datetimetester.py:
```diff
@@ -2501,6 +2501,39 @@ def test_strptime(self):
 
         strptime = self.theclass.strptime
+
+        # bpo-34903: Check that single digit dates and times are allowed.
```
If you have a comment to delim start/end of your block, it means that you have to add a new method, like:

```
def test_strptime_single_digit():
  # bpo-34903: Check that single digit dates and times are allowed.
  ...
```

mikegleen replied:
I'm happy to make this change but note that the code immediately preceding mine has a similar comment for bpo-34482. Could you (@vstinner) and @pganssle please confirm?

vstinner replied:
In that case, move also the previous lines into a subtest (new method).

pganssle replied:
@mikegleen It's fairly common for there to be stylistic and organizational inconsistencies in a file, because the codebase is large, old and has many contributors with different priorities (and different reviewers with different standards).

I tend to agree with Victor here. The main reason I didn't ask for this to be in a new test to start with was because, as you said, the current test suite tends towards these single monolithic tests and I prefer smaller tests and I wasn't sure if monolithic tests was a general policy or not. Given that Victor has much more experience than me with the project's conventions, I'm happy that he asked for this, because it will make me feel more confident in making smaller tests in the future. :smile:

Also, as an aside, I'd like to thank you for your ongoing efforts in this PR, I think we're in the home stretch :wink:

mikegleen committed:
Move code to a separate method.

mikegleen committed:
Merge branch 'master' of https://github.com/mikegleen/cpython

vstinner reviewed (approved):
LGTM.
